### PR TITLE
fix(amazonq): Don't show inline completions when a edit is displayed

### PR DESCRIPTION
**Pull request by floralph**
## Problem
Inline completion suggestions were being shown even when edit suggestions were active, creating a confusing user experience where both types of suggestions could appear simultaneously.

## Solution
1. Added conflict prevention logic in provideInlineCompletionItems() to check for active edit suggestions using the existing aws.amazonq.editSuggestionActive context flag
1. Created isEditSuggestionActive() method to encapsulate context checking
1. Implemented DISCARD telemetry for completion suggestions that can't be shown due to active edits
1. Added unit tests covering the new functionality, including edge cases for mixed suggestion types and items without IDs
1. Updated displayImage.ts to properly set/unset the context flag when edit suggestions are displayed/cleared
1. The solution ensures only one type of suggestion is active at a time while maintaining telemetry compliance and following existing codebase patterns.

---
Treat all work as PUBLIC. Private feature/x branches will not be squash-merged at release time.
Your code changes must meet the guidelines in [CONTRIBUTING.md](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md#guidelines).
License: I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by andrewyuq — APPROVED** [review]

File: packages/amazonq/src/app/inline/completion.ts [hunk 1]
```diff
@@ -435,6 +450,31 @@ ${itemLog}
...
+                                    accepted: false,
+                                    discarded: true,
+                                },
+                            },
+                            firstCompletionDisplayLatency: session.firstCompletionDisplayLatency,
+                            totalSessionDisplayTime: performance.now() - session.requestStartTime,
+                        }
+                        this.languageClient.sendNotification(this.logSessionResultMessageName, params)
+                    }
+                }
```
Comment (andrewyuq) [inline packages/amazonq/src/app/inline/completion.ts]: For UTD we only want to send 1 event for trigger (especially for completion suggestions)
all the items in this completion are discarded we can just send a list of items in completionSessionResult and only one call of `this.languageClient.sendNotification(this.logSessionResultMessageName, params)`

File: packages/amazonq/src/app/inline/completion.ts [hunk 2]
```diff
@@ -237,6 +237,21 @@ export class AmazonQInlineCompletionItemProvider implements InlineCompletionItem
         await vscode.commands.executeCommand(`aws.amazonq.checkInlineSuggestionVisibility`)
     }
 
+    /**
+     * Check if a completion suggestion is currently active/displayed
+     */
+    public isCompletionActive(): boolean {
+        const session = this.sessionManager.getActiveSession()
+        return session !== undefined && session.displayed && !session.suggestions.some((item) => item.isInlineEdit)
```
Comment (leigaol) [inline packages/amazonq/src/app/inline/completion.ts]: `session.displayed` is a flag that indicates if a session has ever been displayed or not. There is an edge case when it is displayed but user did not reject it, instead moved cursor around or kept typing, in that case, the session.display is true but nothing is rendered on screen.

Comment (leigaol) [reply packages/amazonq/src/app/inline/completion.ts]: `await vscode.commands.executeCommand(`aws.amazonq.checkInlineSuggestionVisibility`) ` is the real method to check if anything is rendered on screen or not because this command is defined under a context variable of VS Code and it only runs if inline suggest is visible

File: packages/amazonq/src/app/inline/completion.ts [hunk 3]
```diff
@@ -237,6 +237,21 @@ export class AmazonQInlineCompletionItemProvider implements InlineCompletionItem
...
+    }
+
+    /**
+     * Check if an edit suggestion is currently active
+     */
+    private isEditSuggestionActive(): boolean {
+        return getContext('aws.amazonq.editSuggestionActive') || false
```
Comment (leigaol) [inline packages/amazonq/src/app/inline/completion.ts]: getContext does not necessarily work due to vs code limitations

/**
 * Returns the value of a context key set via {@link setContext} wrapper for this session.
 *
 * Warning: this does not guarantee the state of the context key in vscode because it may have
 * been set via `vscode.commands.executeCommand('setContext')`. It has no connection the
 * context keys stored in vscode itself because an API for this is not exposed.
 */
export function getContext(key: contextKey): any {
    return contextMap[key]
}

File: packages/amazonq/src/app/inline/completion.ts (hunk 1 above)
Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: yep we only need 1 call of logSessionResult per trigger

Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: Updated code to batch completionSessionResult and only send telemetry once.

File: packages/amazonq/src/app/inline/completion.ts (hunk 2 above)
Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: Fixed as suggested.

File: packages/amazonq/src/app/inline/completion.ts (hunk 3 above)
Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: Fixed - The command approach I tried had reliability issues with when clause evaluation. I ended up using a simple boolean property synchronized with setContext calls instead.

File: packages/amazonq/src/app/inline/completion.ts [hunk 4]
```diff
@@ -435,6 +493,14 @@ ${itemLog}
...
             const typeahead = document.getText(new Range(position, editor.selection.active))
 
+            // Check if an edit suggestion is currently active - if so, discard completion suggestions
+            if (this.isEditSuggestionActive()) {
+                this.batchDiscardTelemetryForEditSuggestion(items, session)
+                this.sessionManager.clear()
+                logstr += `- completion suggestions discarded due to active edit suggestion`
+                return []
+            }
+
```
Comment (andrewyuq) [inline packages/amazonq/src/app/inline/completion.ts]: >A newer edit suggestion arriving from the server. In such cases, the currently displayed edit suggestion is hidden from the user with a DISCARD state, and replaced with the newer edit suggestion.

**Review by Will-ShaoHua — DISMISSED** [review]

File: packages/amazonq/src/app/inline/completion.ts (hunk 4 above)
Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: +1, 
```
if (isEditSuggestionActive && !isInlineEdit) {
    return
}
```

Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: the other way i think you can also do 
```
if (!isTriggerByDeletion && !request.partialResultToken && !isEditSuggestionActive) {
    promise.push(completionPromise)
}
```

here in recommendationService https://github.com/aws/aws-toolkit-vscode/blob/master/packages/amazonq/src/app/inline/recommendationService.ts#L135-L141

Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: I think since the only scenario where 2 overlapping suggestion will happen is , correct me if i'm wrong ralph
1. when there is an Edit being shown
2. the Edit is not dismissed by users
3. users again start typing and trigger "Completion"

by doing this, "Completion" is not even fired and you dont even need to consider those suggestion states

File: packages/amazonq/src/app/inline/completion.ts [hunk 5]
```diff
@@ -237,6 +238,63 @@ export class AmazonQInlineCompletionItemProvider implements InlineCompletionItem
...
+            return false
+        }
+
+        // Use VS Code command to check if inline suggestion is actually visible on screen
+        // This command only executes when inlineSuggestionVisible context is true
+        try {
+            await vscode.commands.executeCommand('aws.amazonq.checkInlineSuggestionVisibility')
+            return true
+        } catch {
+            return false
```
Comment (leigaol) [inline packages/amazonq/src/app/inline/completion.ts]: have you managed to get into this catch block?

File: packages/amazonq/src/app/inline/completion.ts (hunk 4 above)
Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: Implemented Will's second solution - added !EditSuggestionState.isEditSuggestionActive() to the condition in recommendationService.ts

Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: we can remove this 
```
  // Check if an edit suggestion is currently active - if so, discard completion suggestions
            if (this.isEditSuggestionActive()) {
                this.batchDiscardTelemetryForEditSuggestion(items, session)
                this.sessionManager.clear()
                logstr += `- completion suggestions discarded due to active edit suggestion`
                return []
            }           
```

Comment (Will-ShaoHua) [reply packages/amazonq/src/app/inline/completion.ts]: ow it will preventing
> A newer edit suggestion arriving from the server. In such cases, the currently displayed edit suggestion is hidden from the user with a DISCARD state, and replaced with the newer edit suggestion.

File: packages/amazonq/src/app/inline/completion.ts (hunk 5 above)
Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: I've implemented your timestamp-based solution using performance.now() in the checkInlineSuggestionVisibility()
method and checking if it was called within the last 50ms.

File: packages/amazonq/src/app/inline/completion.ts (hunk 4 above)
Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts]: Removed

**Review by Will-ShaoHua — APPROVED** [review]

File: packages/amazonq/src/app/inline/completion.ts [hunk 6]
```diff
@@ -237,6 +237,53 @@ export class AmazonQInlineCompletionItemProvider implements InlineCompletionItem
...
+    public async isCompletionActive(): Promise<boolean> {
+        const session = this.sessionManager.getActiveSession()
+        if (session === undefined || !session.displayed || session.suggestions.some((item) => item.isInlineEdit)) {
+            return false
+        }
+
+        // Use VS Code command to check if inline suggestion is actually visible on screen
+        // This command only executes when inlineSuggestionVisible context is true
+        await vscode.commands.executeCommand('aws.amazonq.checkInlineSuggestionVisibility')
+        const isInlineSuggestionVisible = performance.now() - session.lastVisibleTime < 50
```
Comment (andrewyuq) [inline packages/amazonq/src/app/inline/completion.ts:252]: emmm ok

Comment (floralph) [reply packages/amazonq/src/app/inline/completion.ts:252]: This workaround was provided by leigaol@ due to not being able to use getContext in VSC extensions.
